src/watchers/indexable-post.php

```diff
@@ -0,0 +1,322 @@
+<?php
+/**
+ * WordPress Post watcher.
+ *
+ * @package Yoast\YoastSEO\Watchers
+ */
+
+namespace Yoast\YoastSEO\Watchers;
+
+use Yoast\YoastSEO\Exceptions\No_Indexable_Found;
+use Yoast\YoastSEO\WordPress\Integration;
+use Yoast\YoastSEO\Yoast_Model;
+use Yoast\YoastSEO\Models\Indexable;
+
+/**
+ * Fills the Indexable according to Post data.
+ */
+class Indexable_Post implements Integration {
+	/**
+	 * Registers all hooks to WordPress.
+	 *
+	 * @return void
+	 */
+	public function register_hooks() {
+		\add_action( 'wp_insert_post', array( $this, 'save_meta' ), PHP_INT_MAX, 1 );
+		\add_action( 'delete_post', array( $this, 'delete_meta' ) );
+	}
+
+	/**
+	 * Deletes the meta when a post is deleted.
+	 *
+	 * @param int $post_id Post ID.
+	 *
+	 * @return void
+	 */
+	public function delete_meta( $post_id ) {
+		try {
+			$indexable = $this->get_indexable( $post_id, false );
+			$indexable->delete();
+		} catch ( No_Indexable_Found $exception ) {
+			return;
+		}
+	}
+
+	/**
+	 * Saves post meta.
+	 *
+	 * @param int $post_id Post ID.
+	 *
+	 * @return void
+	 */
+	public function save_meta( $post_id ) {
```

Comment:
One thing I'm wondering as I'm reading this is that most of what we're doing here is actually upgrading/migrating/creating posts to the new structure, which just happens to be triggered by a `watcher`.

It might be good to separate this into a different class that does just that and gets triggered by the `watcher`. Especially as in the future we'll want to trigger that very same thing from the frontend and from a user-initiated action in the admin for all posts.